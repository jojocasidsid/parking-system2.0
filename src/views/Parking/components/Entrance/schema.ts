import * as yup from 'yup'

const ERROR_FIELD_REQUIRED = 'This field is required'

export const validationSchema = yup.object({
	vehicle: yup.string().required(ERROR_FIELD_REQUIRED),
	parkedType: yup.number().positive().required(ERROR_FIELD_REQUIRED),
	parkTime: yup.string(),
	parkNow: yup.boolean(),
})

export interface IValidationSchema
	extends yup.InferType<typeof validationSchema> {}

export const defaultValues: IValidationSchema = {
	vehicle: '',
	parkedType: 0,
	parkTime: '',
	parkNow: true,
}
