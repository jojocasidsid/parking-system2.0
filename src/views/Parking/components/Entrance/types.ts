export interface IProps {
	entranceTitle: 'West' | 'East' | 'South'
	slotRefetch: () => void
}
