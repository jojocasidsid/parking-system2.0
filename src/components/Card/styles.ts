import styled from '@emotion/styled'

export const CardWrapper = styled('div')`
	display: flex;
	justify-content: flex-start;
	align-items: center;
	padding: 20px;
	flex-direction: column;
	border-radius: 30px;
	border: 2px solid ${({ theme }) => theme.palette.secondary.main};
`
export const StyledSlotHeader = styled('div')`
	display: flex;
	justify-content: space-around;
	align-items: center;
`
