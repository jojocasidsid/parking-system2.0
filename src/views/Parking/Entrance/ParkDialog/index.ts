import ParkDialog from './ParkDialog'

export default ParkDialog
