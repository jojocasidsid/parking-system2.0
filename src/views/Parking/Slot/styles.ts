import styled from '@emotion/styled'
import DirectionsCarIcon from '@mui/icons-material/DirectionsCar'

export const StyledSlot = styled('div')`
	display: flex;
	justify-content: space-around;
	align-items: center;
`

export const StyledImage = styled(DirectionsCarIcon)`
	font-size: 45px;
`

export const StyledBody = styled('div')`
	display: flex;
	justify-content: center;
	align-items: center;
	flex-direction: column;
	text-align: center;
`
