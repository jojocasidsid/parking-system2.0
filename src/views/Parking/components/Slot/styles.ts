import styled from '@emotion/styled'
import DirectionsCarIcon from '@mui/icons-material/DirectionsCar'

export const ThumbnailWrapper = styled('div')`
	display: flex;
	justify-content: flex-start;
	align-items: center;
	padding: 20px;
	flex-direction: column;
	border-radius: 30px;
	border: 2px solid ${({ theme }) => theme.palette.secondary.main};
`
export const StyledSlot = styled('div')`
	display: flex;
	justify-content: space-around;
	align-items: center;
`

export const StyledImage = styled(DirectionsCarIcon)`
	font-size: 45px;
`

export const ActionWrapper = styled('div')`
	& > button {
		margin: 20px;
	}
`
