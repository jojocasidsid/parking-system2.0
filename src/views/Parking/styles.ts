import styled from '@emotion/styled'

export const StyledRoot = styled('div')`
	display: flex;
	justify-content: center;
	align-items: center;
	flex-direction: column;
	padding: 20px;
`

export const ParkingWrapper = styled('div')`
	display: flex;
	justify-content: flex-start;
	align-items: center;
	padding: 20px;
	flex-wrap: nowrap;
`

export const SlotsWrapper = styled('div')`
	display: grid;
	grid-template-columns: repeat(5, 1fr);
	grid-template-rows: repeat(5, 1fr);
	grid-gap: 10px;
	width: 90vw;
`
