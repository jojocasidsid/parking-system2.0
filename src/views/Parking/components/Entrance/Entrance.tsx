import React, { useState } from 'react'
import { Button, Typography } from '@mui/material'

import { useSnackbar } from 'notistack'
import SlotsAPI from 'apis/Slots/SlotsApi'
import { yupResolver } from '@hookform/resolvers/yup'
import { useForm } from 'react-hook-form'
import { IProps } from './types'
import ParkDialog from './ParkDialog'
import { defaultValues, validationSchema, IValidationSchema } from './schema'

const Entrance = ({ entranceTitle, slotRefetch }: IProps) => {
	const [open, setIsOpen] = useState(false)
	const [submitting, setSubmitting] = useState(false)
	const { enqueueSnackbar } = useSnackbar()

	const handleClose = () => {
		setIsOpen(false)
	}

	const handleOpen = () => {
		setIsOpen(true)
	}

	const {
		handleSubmit,
		control,
		reset,
		watch,
		formState: { errors },
	} = useForm({
		mode: 'all',
		reValidateMode: 'onChange',
		defaultValues,
		resolver: yupResolver(validationSchema),
	})

	const onSubmit = async (data: IValidationSchema) => {
		setSubmitting(true)
		const { parkedType, vehicle, parkNow, parkTime } = data
		try {
			// get neareast slot
			const getParking = await SlotsAPI.getNearest(parkedType, entranceTitle)

			// throw error if no slot available
			if (!getParking) {
				throw new Error('There is no available slot')
			}

			// park on that nearest slot
			const nearestParkingSlotId = getParking.data[0].id
			await SlotsAPI.parkSlot(
				nearestParkingSlotId,
				parkedType,
				vehicle,
				parkNow || false,
				parkTime || ''
			)

			enqueueSnackbar('Vehicle has been successfully parked.', {
				variant: 'success',
			})

			slotRefetch()
		} catch (e) {
			enqueueSnackbar(`Encountered an error`, { variant: 'error' })
		}
		setSubmitting(false)
		handleClose()
		reset()
	}

	return (
		<div>
			<ParkDialog
				open={open}
				handleClose={handleClose}
				entranceTitle={entranceTitle}
				onSubmit={onSubmit}
				submitting={submitting}
				handleSubmit={handleSubmit}
				control={control}
				errors={errors}
				watch={watch}
				reset={reset}
			/>
			<Button onClick={() => handleOpen()}>
				<Typography variant='h3'>Park ({entranceTitle})</Typography>
			</Button>
		</div>
	)
}

export default Entrance
