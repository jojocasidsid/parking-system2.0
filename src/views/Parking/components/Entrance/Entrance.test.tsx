import React from 'react'
import renderer from 'react-test-renderer'
import { QueryClient, QueryClientProvider } from 'react-query'
import { SnackbarProvider } from 'notistack'
import { ThemeProvider } from '@mui/material/styles'
// config
import theme from 'config/theme'

// component
import Entrance from './Entrance'

const queryClient = new QueryClient()

describe('Entrance Component', () => {
	it('should render east entrance component without crashing', () => {
		const tree = renderer.create(
			<QueryClientProvider client={queryClient}>
				<ThemeProvider theme={theme}>
					<SnackbarProvider
						maxSnack={5}
						anchorOrigin={{
							vertical: 'bottom',
							horizontal: 'right',
						}}
						preventDuplicate
						autoHideDuration={5000}
					>
						<Entrance entranceTitle='East' slotRefetch={() => {}} />
					</SnackbarProvider>
				</ThemeProvider>
			</QueryClientProvider>
		)

		expect(tree).toMatchSnapshot()
	})

	it('should render west entrance component without crashing', () => {
		const tree = renderer.create(
			<QueryClientProvider client={queryClient}>
				<ThemeProvider theme={theme}>
					<SnackbarProvider
						maxSnack={5}
						anchorOrigin={{
							vertical: 'bottom',
							horizontal: 'right',
						}}
						preventDuplicate
						autoHideDuration={5000}
					>
						<Entrance entranceTitle='West' slotRefetch={() => {}} />
					</SnackbarProvider>
				</ThemeProvider>
			</QueryClientProvider>
		)

		expect(tree).toMatchSnapshot()
	})

	it('should render south entrance component without crashing', () => {
		const tree = renderer.create(
			<QueryClientProvider client={queryClient}>
				<ThemeProvider theme={theme}>
					<SnackbarProvider
						maxSnack={5}
						anchorOrigin={{
							vertical: 'bottom',
							horizontal: 'right',
						}}
						preventDuplicate
						autoHideDuration={5000}
					>
						<Entrance entranceTitle='South' slotRefetch={() => {}} />
					</SnackbarProvider>
				</ThemeProvider>
			</QueryClientProvider>
		)

		expect(tree).toMatchSnapshot()
	})
})
