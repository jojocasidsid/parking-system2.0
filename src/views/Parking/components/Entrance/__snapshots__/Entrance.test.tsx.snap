// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entrance Component should render east entrance component without crashing 1`] = `
<div>
  <button
    className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1k8fnjm-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <h3
      className="MuiTypography-root MuiTypography-h3 css-1dscty8-MuiTypography-root"
    >
      Park (
      East
      )
    </h3>
  </button>
</div>
`;

exports[`Entrance Component should render south entrance component without crashing 1`] = `
<div>
  <button
    className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1k8fnjm-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <h3
      className="MuiTypography-root MuiTypography-h3 css-1dscty8-MuiTypography-root"
    >
      Park (
      South
      )
    </h3>
  </button>
</div>
`;

exports[`Entrance Component should render west entrance component without crashing 1`] = `
<div>
  <button
    className="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1k8fnjm-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <h3
      className="MuiTypography-root MuiTypography-h3 css-1dscty8-MuiTypography-root"
    >
      Park (
      West
      )
    </h3>
    <span
      className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
